File: 0752/solution.py
from collections import deque


class Solution:
    def openLock(self, deadends: list[str], target: str) -> int:
        if target == "0000":
            return 0

        def lock_converter(choice):
            a = set()
            b = set()
            c = a
            res = []
            nums = [-1, 1]
            for i in range(len(choice)):
                for num in nums:
                    out = (int(choice[i]) + num) % 10
                    res.append(choice[:i] + str(out) + choice[i + 1 :])
            return res

        bfs = deque()
        visited = set(deadends)
        if ("0000") in visited:
            return -1
        bfs.append(["0000", 0])
        visited.add("0000")
        while bfs:
            key, moves = bfs.popleft()
            if key == target:
                return moves
            children = lock_converter(key)
            for child in children:
                if child not in visited:
                    visited.add(child)
                    bfs.append([child, moves + 1])

        return -1

File: 0752/solution.ts
const combinations = (choice: string): string[] => {
  const res: string[] = [];
  for (let i = 0; i < choice.length; i++) {
    const num = Number(choice[i]);
    const small = (num + 1) % 10;
    const big = num == 0 ? 9 : num - 1;
    res.push(choice.substring(0, i) + String(small) + choice.substring(i + 1));
    res.push(choice.substring(0, i) + String(big) + choice.substring(i + 1));
  }
  return res;
};

function openLock(deadends: string[], target: string): number {
  if (target === "0000") return 0;
  const visited = new Set<string>(deadends);
  const bfs: [string, number][] = [["0000", 0]];
  if (visited.has("0000")) {
    return -1;
  }
  while (bfs.length) {
    const [choice, moves] = bfs.shift()!;
    if (choice === target) {
      return moves;
    }
    const children = combinations(choice);
    for (let child of children) {
      if (!visited.has(child)) {
        visited.add(child);
        bfs.push([child, moves + 1]);
      }
    }
  }
  return -1;
}
